Feat/Scrum 151 노트 작성, 곡 검색 기능 구현

derrickkim0109 opened the pull request:
## PR 타입
어떤 변경에 대한 PR인가요?

<!-- 아래 체크리스트 중 해당 되는 부분에 체크해주세요 "x". -->

- [ ] 버그 수정
- [x] Feature / 신기능
- [ ] 코드 스타일 업데이트 (formatting, local variables)
- [ ] 리팩토링 (no functional changes, no api changes)
- [ ] 빌드 관련 변경사항
- [ ] CI 관련된 변경사항
- [ ] Documentation content changes
- [ ] 구조 변경
- [ ] Other... 추가 설명 요망:

## 배경

<!-- 해당 PR에서 개발한 내용에 대해 알아야 하는 배경지식을 작성해주세요. -->

- 노트 작성에 필요한 모든 객체를 구현해야 합니다. 

## 목표

<!-- 해당 PR에서 개발/수정 한 기능의 목표를 작성해주세요. -->

- 노트 작성 UI
  - [x] 완료 버튼
    - [x]  곡 추가 & 노트 TextView(:생각을 남겨보세요가 있는 텍스트 뷰) 텍스트 입력 시 버튼 활성화
    - [x] Optional하게 가사, 가사 배경을 PostNote API로 call 
  
  - [x] 곡 추가
    - [x] 곡 검색 ViewController로 이동 
    - [x] 곡 선택 후, item 넘기는 Delegate -> Coordinator로 구현
    
  - [x] 가사 입력하는 텍스트 뷰 
    - [x] 곡 추가가 되기 전까지는 입력 불가 Alert
    - [x] 가사 50자까지 입력가능
    - [x] 텍스트 상시 정 중앙 위치
    - [x] 텍스트 카운트 Label 오른쪽 아래에 위치 

    - [x] 가사 배경 버튼
      - [x] 가사 텍스트 뷰 입력 전까지 액션 불가 Alert 
      - [x] 가사 배경 선택할 수 있는 LyricsBackground ViewController로 이동
        - [x] AnyPublisher로 데이터 이동

      - [x] 가사 검색
        - [x] 웹뷰 URL 네이버 search로 가사 검색

      - [x] 노트 작성 TextView 
        - [x] 텍스트 카운트 Label 
            - [x] 오른쪽 아래 위치
            - [x] 키보드 Height에 맞춰서 위치 위/아래로 이동

## 결과 (Optional)

<!-- 개발한 내용의 결과를 설명해주세요. !-->

|노트작성|곡 선택 x > 가사 작성|모두 작성|
|--|--|--|
|<img src="https://github.com/user-attachments/assets/9120a99a-60c9-4401-ab3e-1d67a637ea1d" width="250">|<img src="https://github.com/user-attachments/assets/d07e3c2c-1927-4726-99e7-2e71c5600c54" width="250">|<img src="https://github.com/user-attachments/assets/21ba3dec-3a16-43b5-bda1-2ec4b6d8eff7" width="250">|

|가사배경 선택| 가사 검색|
|--|--|
|<img src="https://github.com/user-attachments/assets/ca8755b0-eb5f-4cec-8eec-e5e84220f231" width="250">|<img src="https://github.com/user-attachments/assets/014f9052-8f05-4040-ba73-5871bb9a4818" width="250">|

| 곡검색 - 기본| 곡검색 - 키워드 입력|
|--|--|
|<img src="https://github.com/user-attachments/assets/7c3cfc2b-b429-417e-822d-a2e09b3ca8d6" width="250">|<img src="https://github.com/user-attachments/assets/ef596134-acde-4b72-9912-222b9a0dba0a" width="250">|

derrickkim0109 committed:
feat: FeelinAPI postNote 추가

derrickkim0109 committed:
feat<LyricsBackground>: 가사 배경 선택을 위한 ViewController 구현

derrickkim0109 committed:
style: 각 Coordinator 네이밍 변경

derrickkim0109 committed:
feat: Search Song WebViewController 구현

derrickkim0109 committed:
feat: Search Songs ViewController 구현

derrickkim0109 committed:
refactor: 공용 컴포넌트 NavigationBar 수정

derrickkim0109 committed:
refactor: PaginationService으로 네이밍 변경 및 DomainShared로 위치 이동

derrickkim0109 committed:
add: CocoaCombine Extension 추가

- UIGesture
- TextView
- UIView

derrickkim0109 committed:
refactor: ProfileViewModel 불필요한 코드 수정

derrickkim0109 committed:
refactor: 각 버튼 인터렉션에 따른 UI update로직 추가

- PostSelect
- FeelinConfirm
- FeelinSelectableImage

derrickkim0109 committed:
add: Feelin API search song api 추가

derrickkim0109 committed:
feat: Post Note를 위한 UseCase 추가

derrickkim0109 committed:
feat: Search Song UseCase 구현

derrickkim0109 committed:
feat: NoteService 구현

derrickkim0109 committed:
feat: PostNoteView navigation Bar 추가

derrickkim0109 committed:
feat: Search Song API 추가

derrickkim0109 committed:
refactor: 각 PaginationService 이름 변경

derrickkim0109 committed:
refactor: NavigationBar 불필요한 패딩값 제거

derrickkim0109 committed:
chore: 파일 이름 주석 변경

derrickkim0109 committed:
feat: UITextView shouldBeginEditingSubject Combine 기능 구현

derrickkim0109 committed:
remove: 불필요한 뷰 제거

derrickkim0109 committed:
feat: UITextView 라인 수 구하는 로직 구현

derrickkim0109 committed:
refactor: Note API 변경으로 인해 매개변수 변경

derrickkim0109 committed:
refactor: PostNoteUseCase Protocol 추가

derrickkim0109 committed:
refactor: Pagination 이름 변경으로 DI 내부 프로퍼티 이름 변경

derrickkim0109 committed:
feat: MainCoordinatorInterface PostNote ViewController 추가

derrickkim0109 committed:
feat: Feature Main 모듈에 domain 추가

derrickkim0109 committed:
feat: Search WebView 구현

derrickkim0109 committed:
refactor: Lyrics Background Publisher 네이밍, 타입 변경

derrickkim0109 committed:
refactor: SearchSongView 컬렉션 뷰 높이 및 셀 사이즈 조정

derrickkim0109 committed:
feat: SearchSong ViewController 및 ViewModel 구현

derrickkim0109 committed:
chore: 불필요한 설정 제거

derrickkim0109 committed:
refactor: PostSelectButton UI 업데이트 메서드 수정

derrickkim0109 committed:
feat: PostNoteview 키보드 높이에 맞게 noteCountLabel 위치 변경

- view 컴포넌트 업데이트 메서드 추가
- private 처리
- 불필요한 설정 제거

derrickkim0109 committed:
feat: PostNote ViewController 및 ViewModel 구현

derrickkim0109 committed:
feat: SongsError 구현

derrickkim0109 committed:
test: FeatureMain Example 노트작성 부분 추가

derrickkim0109 committed:
refactor: 불필요한 코드 및 빠진 설정 추가

inwoodev commented on Projects/Feature/Main/Interface/Sources/Views/SearchBarView.swift:
FeelinSearchBar가 있긴한데요~ 해당 UI를 이용하기에는 UI적으로 차이가 있으셨을까요~?

derrickkim0109 replied:
이 부분은 제거했는데 커밋에 남아서 올라간 것 같습니다. Outdated로 표기되고 있어요.

inwoodev commented on Projects/Shared/Util/Sources/Extensions/Combine/UIGestureRecognizer+Combine.swift:
👍👍👍

inwoodev commented on Projects/Shared/DesignSystem/Sources/Buttons/FeelinConfirmButton.swift:
요 변경 부분에 대한 히스토리가 궁금합니다~

derrickkim0109 replied:
PostNoteView에서 FeelinConfirmButton를 사용할 때, isEnabled의 상태값에 따라 color를 변경할 필요가 있었습니다. 혹시 사용하시는 곳에서 기존 설정이 필요한 곳이 있을까요? 
그렇다면 제가 새로 버튼을 만들어야 할 것 같아서요.

inwoodev replied:
오 아뇨아뇨. 변경사항 유지하셔도 괜찮을듯합니다~~

inwoodev commented on Projects/Shared/Util/Sources/Extensions/UITextView+.swift:
```diff
@@ -0,0 +1,26 @@
+//  Created by Derrick kim on 9/6/24.
+//
+
+import UIKit
+
+public extension UITextView {
+    func numberOfLine() -> Int {
+        let layoutManager = self.layoutManager
+
+        var numberOfLines = 0
+        var index = 0
+
+        while index < layoutManager.numberOfGlyphs {
+            var range = NSRange(location: 0, length: 0)
+            layoutManager.lineFragmentRect(forGlyphAt: index, effectiveRange: &range)
+            numberOfLines += 1
+            index = NSMaxRange(range)
+        }
+
+        return numberOfLines
+    }
+}
```
👍👍👍

inwoodev commented on Projects/Feature/Main/Interface/Sources/Views/PostNoteView.swift:
```diff
@@ -315,4 +328,69 @@ final class PostNoteView: UIView {
+        UIView.animate(withDuration: 0.3) {
+            self.layoutIfNeeded()
+        }
+    }
```
이미 구현되어있는 CombineKeyboard를 사용하여 구현하기에는 무리였다 판단하시고 추가적으로 구현하신걸까요~?

derrickkim0109 replied:
이 부분은 빼먹었네요. 

다시 수정해서 올리겠습니다~!

inwoodev commented on Projects/Feature/Main/Interface/Sources/ViewControllers/PostNoteViewController.swift:
```diff
@@ -42,128 +56,280 @@ public final class PostNoteViewController: UIViewController {
+
+        selectLyricsBackgroundButton.tapPublisher
+            .receive(on: DispatchQueue.main)
+            .flatMap { [weak self] _ -> AnyPublisher<Void, Never> in
+                guard let self = self else { return Empty().eraseToAnyPublisher() }
+
+                if !self.selectLyricsBackgroundButton.isEnabled {
+                    self.showAlert(
+                        title: "곡을 추가한 후,\n가사배경을 추가할 수 있어요.",
+                        message: nil,
+                        singleActionTitle: "확인"
+                    )
+                    return Empty().eraseToAnyPublisher()
+                }
+
+                return Just(()).eraseToAnyPublisher()
+            }
```
요기서 굳이 flatMap을 사용해야할까 하는 생각이 있긴 합니다~
p4

derrickkim0109 replied:
수정했습니다~!

inwoodev commented on Projects/Feature/Main/Interface/Sources/ViewModels/PostNoteViewModel.swift:
```diff
@@ -5,21 +5,143 @@
+        }
+        .eraseToAnyPublisher()
+
+        // 필수 필드와 선택 필드를 결합
+        let combinedPublisher = requiredFieldsPublisher
+            .combineLatest(optionalFieldsPublisher)
+            .map { (requiredFields, optionalFields) -> PostNoteValue in
+                let (song, noteContent, status) = requiredFields
+                let (lyrics, lyricsBackground) = optionalFields
+
+                return PostNoteValue(
+                    id: song.id,
+                    lyrics: lyrics,
+                    background: lyricsBackground,
+                    content: noteContent,
+                    status: status
+                )
+            }
+            .eraseToAnyPublisher()
+
+        // 완료 버튼 탭과 결합된 필드로 노트를 게시
+        return input.completeButtonTapPublisher
+            .combineLatest(combinedPublisher)
+            .flatMap { (_, value) in
+                return self.postNote(value)
+            }
+            .eraseToAnyPublisher()
+
+    }
+
+    func postNote(_ requestValue: PostNoteValue) -> AnyPublisher<NoteResult, Never> {
+        return self.postNoteUseCase
+            .execute(value: requestValue)
+            .mapError(NoteError.init)
+            .receive(on: DispatchQueue.main)
+            .catch { error in
+                return Just(.failure(error))
+                    .eraseToAnyPublisher()
+            }
+            .eraseToAnyPublisher()
```
깔쌈하네요~ 다만 UseCase가 외부에서 주입되기 되기도 하며 순환참조를 막기 위해서 129번째 줄은 `[weak self]`로 가는게 어떤가 싶어요~
p4

inwoodev replied:
아 그리고 NoteResult가 하나 보이긴 한데. 제가 추가한 `.mapToResult()`도 한번 확인 해 보시면 중복 코드를 줄일 수 있지 않나 싶습니다:)

derrickkim0109 replied:
수정했습니다~!

inwoodev commented on Projects/Feature/Main/Interface/Sources/ViewControllers/SearchSongViewController.swift:
유저가 검색 중에 화면을 스크롤하게 된다면 가장 하단 cell은 키보드에 가려서 확인을 못할 거 같은데. 어떻게 생각하시나요~?

CombineKeyboard와 같이 키보드 높이를 체크하는 로직 활용을 고려 해 볼 수 있지 않나 싶습니다:)

p3

derrickkim0109 replied:
이 부분은 인우님이 사용하셨던 방법과 달리 noteCountLabel은 NSLayoutConstraint으로 Autolayout이 잡혀있어 어려울 것 같습니다.

inwoodev replied:
그럼에도 불구하고 하단 cell이 키보드에 의해 가려지면 사용성 측면에서 별로일거 같아요. CombineKeyboard가 안된다면 다른 방법을 찾아봐야하지 않을까요~?

derrickkim0109 replied:
음.. 제가 말한 부분은 인우님이 말한 부분과는 다른 부분이였습니다. 

noteCountLabel은 곡 검색에 있는 부분이 아니라 위에서 PostNoteView에서 CombineKeyboard를 언급한 부분이라 커뮤니케이션 미스가 있었던 것 같습니다. 

곡 검색 부분은 CombineKeyboard로 체크하는 방식으로 변경했습니다.

inwoodev commented on Projects/Domain/Note/Interface/Sources/Services/NoteServiceInterface.swift:
이미 `NoteAPIService`가 있는데 요걸 활용하는건 어떤가요~? 추가적으로 만들어야할 이유가 있는지 궁금합니다~!
p3

derrickkim0109 replied:
말씀하신 부분으로 다시 수정해두겠습니다~!

inwoodev commented on Projects/Feature/Main/Interface/Sources/Views/SearchSongView.swift:
```diff
@@ -70,6 +69,10 @@ final class SearchSongView: UIView {
     override func layoutSubviews() {
         super.layoutSubviews()
 
+        if let layout = collectionView.collectionViewLayout as? UICollectionViewFlowLayout {
+            layout.itemSize = CGSize(width: collectionView.bounds.width, height: 64)
+        }
+
```
요 부분에 대해서 간단하게 설명 해 주실 수 있나요~?

derrickkim0109 replied:
이 부분은 collectionview의 아이템의 사이즈를 정해주는 부분이 레이아웃 에러 발생해서 layoutSubviews 내부에 선언했습니다.

혹시 더 좋은 방법이 있을까요?

inwoodev replied:
다른 방법으로는 TableView를 사용하는 방법도 있긴 합니다만 지금 이 방식도 문제 없어보입니다~

inwoodev commented on Projects/Domain/Note/Interface/Sources/Model/NoteResult.swift:
```diff
@@ -0,0 +1,13 @@
+
+import Foundation
+
+public enum NoteResult {
+    case success
+    case failure(NoteError)
+}
```
`.mapToResult()`를 활용하면 따로 필요 없을 듯 해요!!

derrickkim0109 replied:
수정했습니다~!

derrickkim0109 replied:
삭제했습니다~!

inwoodev commented on Projects/Domain/Shared/Interface/Sources/Models/PaginationService.swift:
이 부분이 왜 한 번 더 추가가 되었는지 흠...이상하네요~~ :)

derrickkim0109 replied:
음.. rebase하는 과정에서 복사가 되서 다시 rebase했는데.. 
제거해두겠습니다.

derrickkim0109 replied:
추가적으로 이 부분은 이미 마지막 푸시에서 제거된 상태라 따로 수정할 수 있는 부분이 없을 것 같습니다.

derrickkim0109 committed:
chore: 불필요한 코드 제거 및 스코프 추가

inwoodev commented on Projects/Feature/Main/Interface/Sources/Views/PostNoteView.swift:
```diff
@@ -40,18 +69,20 @@ final class PostNoteView: UIView {
+        label.text = "실리카겔"
 
         return label
     }()
```
76, 85에 있는 원시값들은 이제 필요가 없어 보여요!
p3

derrickkim0109 replied:
수정하겠습니다~!

inwoodev reviewed (commented):
고생하였습니다~ 코멘트 몇 자 남겨 봤어요. 확인 해 주세요~

derrickkim0109 committed:
remove: 불필요한 파일제거

derrickkim0109 committed:
refactor: 불필요한 초기 설정 제거

derrickkim0109 committed:
refactor: 불필요한 객체 제거

- NoteResult
- NoteService -> NoteAPIService로 대체
- Source 폴더 내부에 NoteAPIService로 extension 메서드 구현

derrickkim0109 committed:
refactor: Domain Note 모든 UseCase UseCases 폴더로 이동

derrickkim0109 committed:
refactor: PostNoteViewModel 내부 메서드 mapToResult로 변경

derrickkim0109 committed:
refactor: weak self로 UseCase 메서드 변경

derrickkim0109 committed:
refactor: MainCoordinator Interface NoteAPI Service로 변경

derrickkim0109 committed:
refactor: 가사검색 WebView 기획 변경으로 인해 Melon으로 사이트 변경

derrickkim0109 committed:
refactor: 빈 텍스트에도 label의 사이즈가 잡히도록 수정

derrickkim0109 committed:
fix: 곡 검색 시 키보드에 리스트 가리는 부분 수정

derrickkim0109 committed:
refactor: Search Song List flowLayout.itemSize 설정 분리

inwoodev commented on Projects/Feature/Main/Interface/Sources/ViewControllers/SearchSongWebViewController.swift:
빠른 수정 👏👏👏👏 고생하셨습니다~

inwoodev commented on Projects/Feature/Main/Interface/Sources/Views/SearchSongView.swift:
```diff
@@ -104,4 +102,8 @@ final class SearchSongView: UIView {
                     .grow(1)
             }
     }
+
+    private func setUpSongsCollectionViewLayout() {
+        flowLayout.itemSize = CGSize(width: collectionView.frame.width, height: 64)
+    }
```
이게 조금 더 깔끔한듯 하네요 👍👍👍

inwoodev commented on Projects/Feature/Main/Interface/Sources/ViewModels/PostNoteViewModel.swift:
```diff
@@ -133,15 +136,11 @@ private extension PostNoteViewModel {
+            .mapError(NoteError.init)
+            .mapToResult()
     }
 }
```
💪💪💪

inwoodev reviewed (approved):
고생하셨습니다~ 추가 수정사항이 없으시다면 머지하셔도 좋습니다!!
